### ISSUE TITLE
Enable linting with phpcs-changed locally

### ISSUE THREAD

**Issue by coreymckrill** [woothemes/woocommerce#36850]
> **Note**
> **TL;DR** is there a way we could install and run phpcs-changed from the root of the Git repo instead of only from `plugins/woocommerce`?

Right now phpcs-changed is run as part of a GitHub workflow whenever you push changes to a pull request, which is great because it only shows you linting errors on the lines you changed/added, instead of in each entire file. However, it'd be even better if I could run this command locally before I push changes up to GitHub, so I don't have to wait for all the checks to finish to know if I have linting errors.

In `plugins/woocommerce/composer.json` I see the `phpcs-pre-commit` [script](https://github.com/woocommerce/woocommerce/blob/e05a191156aa9c59d8bf94429fba4dbc1a92519e/plugins/woocommerce/composer.json#L91-L93), but A) it doesn't seem to actually run when I attempt a commit, and B) the command doesn't work on it's own as it currently is.

I have tried running phpcs-changed manually a few different ways, and there does not appear to be any way to successfully use the instance that gets installed and managed by composer in the `plugins/woocommerce` directory. What happens is, with the right command line arguments, it _can_ be run, but it shows you every linting error in each file instead of just on the changed lines.

Example command, run from the `plugins/woocommerce` directory (this is to compare a PR branch to trunk, where everything has already been committed):
```bash
composer exec phpcs-changed -- -s --git --git-base trunk $(git diff $(git merge-base HEAD trunk) --name-only -- '*.php')
```

Eventually I discovered that the issue is that you have to run the command from the directory where the composer.json file is located (`plugins/woocommerce`), but the phpcs-changed command assumes that you are in the root of the Git repository, so it can't find the files that you're specifying when it's getting the linting results from the base branch. Thus it assumes that all of the changed files are brand new and shows you every linting error.

The workaround I'm using right now is to install phpcs-changed and some other necessary packages globally and then I can run it from the Git repo root without having to go through composer. However, this is not ideal because:
* It makes the monorepo less self-contained, and takes us farther away from the goal of having everything ready to go after just running the install command.
* I think it would actually be great if phpcs-changed could be run on the pre-commit hook.

**Comment by sirbrillig** [comment]
Does it work if you run phpcs-changed from the git root directory by specifying a path instead of using `composer exec`?

```
./plugins/woocommerce/vendor/bin/phpcs-changed -s --git --git-base trunk $(git diff $(git merge-base HEAD trunk) --name-only 'file.php')
```

> the phpcs-changed command assumes that you are in the root of the Git repository

I mentioned it in the other issue, but this should not be the case. `phpcs-changed` automatically detects the git root directory and should be able to operate from anywhere. It's likely I'm misunderstanding the problem, though.

**Comment by coreymckrill** [comment]
Thanks for taking a look, @sirbrillig!

> Does it work if you run phpcs-changed from the git root directory by specifying a path instead of using composer exec?

If I call the phpcs-changed script directly, outside of the composer context, the other necessary composer dependencies (phpcs, wp-coding-standards) aren't available. I get `ERROR: Cannot find executable for phpcs` unless I already have those dependencies installed globally.

> phpcs-changed automatically detects the git root directory and should be able to operate from anywhere.

I'm probably not totally understanding what's happening either, but here's an example:

[This PR](https://github.com/woocommerce/woocommerce/pull/37303) is a one-liner for the file `templates/cart/cart.php` (relative to plugins/woocommerce, not the git root). If I run this command:

```
composer exec phpcs-changed -- -s --git --git-base trunk $(git diff $(git merge-base HEAD trunk) --name-only -- '*.php')
```

I get `ERROR: Cannot read file 'plugins/woocommerce/templates/cart/cart.php'` because the list of changed files produced by git is relative to the git root, not the current directory.

Ok, so I add `--relative` to the command so that the list of changed files will be relative to the current directory:

```
composer exec phpcs-changed -- -s --git --git-base trunk $(git diff $(git merge-base HEAD trunk) --relative --name-only -- '*.php')
```

Now the command works, but it's giving me 25 linting errors (for the whole file) instead of one for the one line that changed. If I run the same command again with `--debug` I see this (filtered for relevant lines):

```
checking git existence of file with command:
git status --porcelain 'templates/cart/cart.php'
git status output:


is run from git root: false
running modified file phpcs command:
git show HEAD:$(git ls-files --full-name 'templates/cart/cart.php') | phpcs --report=json -q --stdin-path='templates/cart/cart.php' -
modified file phpcs command output:
{a bunch of JSON representing the 25 errors}

checking status of file with command:
git cat-file -e 'c5564a15c11eae5b30de7ab6e37edbd973a29ebe':'templates/cart/cart.php' 2>/dev/null
status command output:


status command return val:
128
Skipping the linting of the unmodified file as it is a new file.
```

(c5564a15c11eae5b30de7ab6e37edbd973a29ebe in this case is the hash for the commit on trunk that the PR branch split off from)

So the git command phpcs-changed is using can't find the file in trunk, so it considers it a new file and thus gives me all the errors.

**Comment by sirbrillig** [comment]
Thank you for the info! It sounds like for some reason the command `git cat-file -e 'c5564a15c11eae5b30de7ab6e37edbd973a29ebe':'templates/cart/cart.php'` is not finding the file, which is weird since [the file definitely exists at that version](https://github.com/woocommerce/woocommerce/blob/c5564a15c11eae5b30de7ab6e37edbd973a29ebe/plugins/woocommerce/templates/cart/cart.php).

Indeed, I tried checking out the branch myself and I get the same result:

```
> cd plugins/woocommerce
> git cat-file -e 'c5564a15c11eae5b30de7ab6e37edbd973a29ebe':'templates/cart/cart.php'
fatal: path 'templates/cart/cart.php' does not exist in 'c5564a15c11eae5b30de7ab6e37edbd973a29ebe'
> echo $?
128
```

However, I noticed that in other commands I use the full path to the file rather than the relative path, so I tried that for this command:

```
> cd plugins/woocommerce
> git cat-file -e 'c5564a15c11eae5b30de7ab6e37edbd973a29ebe':$(git ls-files --full-name 'templates/cart/cart.php')
> echo $?
0
```

It worked! 

I think that I need to update `phpcs-changed` to always use the full path for its operations. I'll work on this in https://github.com/sirbrillig/phpcs-changed/issues/66 and let you know when it's fixed.
